File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter17/FingerPaint/FingerPaint/BitmapPrintDocument.cs
using System;
using Windows.Graphics.Printing;
using Windows.Graphics.Printing.OptionDetails;
using Windows.UI.Core;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Media.Imaging;
using Windows.UI.Xaml.Printing;

namespace FingerPaint
{
    public class BitmapPrintDocument : PrintDocument
    {
        Func<BitmapSource> getBitmap;
        IPrintDocumentSource printDocumentSource;

        // Element to print
        Border border = new Border
        {
            Child = new Image()
        };

        public BitmapPrintDocument(Func<BitmapSource> getBitmap)
        {
            this.getBitmap = getBitmap;

            // Get IPrintDocumentSource and attach event handlers
            printDocumentSource = this.DocumentSource;
            this.Paginate += OnPaginate;
            this.GetPreviewPage += OnGetPreviewPage;
            this.AddPages += OnAddPages;

            // Attach PrintManager handler
            PrintManager.GetForCurrentView().PrintTaskRequested += 
                                                OnPrintDocumentPrintTaskRequested;
        }

        async void OnPrintDocumentPrintTaskRequested(PrintManager sender, 
                                                     PrintTaskRequestedEventArgs args)
        {
            PrintTaskRequestedDeferral deferral = args.Request.GetDeferral();

            // Obtain PrintTask
            PrintTask printTask = args.Request.CreatePrintTask("Finger Paint", 
                                                               OnPrintTaskSourceRequested);

            // Probably set orientation to landscape
            PrintTaskOptionDetails optionDetails = 
                PrintTaskOptionDetails.GetFromPrintTaskOptions(printTask.Options);

            PrintOrientationOptionDetails orientation = 
                optionDetails.Options[StandardPrintTaskOptions.Orientation] as 
                                                            PrintOrientationOptionDetails;

            bool bitmapIsLandscape = false;

            await border.Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
                {
                    BitmapSource bitmapSource = getBitmap();
                    bitmapIsLandscape = bitmapSource.PixelWidth > bitmapSource.PixelHeight;
                });

            orientation.TrySetValue(bitmapIsLandscape ? PrintOrientation.Landscape : 
                                                        PrintOrientation.Portrait);

            deferral.Complete();
        }

        void OnPrintTaskSourceRequested(PrintTaskSourceRequestedArgs args)
        {
            args.SetSource(printDocumentSource);
        }

        void OnPaginate(object sender, PaginateEventArgs args)
        {
            PrintPageDescription pageDesc = args.PrintTaskOptions.GetPageDescription(0);

            // Get the Bitmap
            (border.Child as Image).Source = getBitmap();

            // Set Padding on the Border
            double left = pageDesc.ImageableRect.Left;
            double top = pageDesc.ImageableRect.Top;
            double right = pageDesc.PageSize.Width - left - pageDesc.ImageableRect.Width;
            double bottom = pageDesc.PageSize.Height - top - pageDesc.ImageableRect.Height;
            border.Padding = new Thickness(left, top, right, bottom);
            
            this.SetPreviewPageCount(1, PreviewPageCountType.Final);
        }

        void OnGetPreviewPage(object sender, GetPreviewPageEventArgs args)
        {
            this.SetPreviewPage(args.PageNumber, border);            
        }

        void OnAddPages(object sender, AddPagesEventArgs args)
        {
            this.AddPage(border);
            this.AddPagesComplete();
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter17/FingerPaint/FingerPaint/BitmapPrintDocument.cpp

#include "pch.h"
#include "BitmapPrintDocument.h"

using namespace FingerPaint;

using namespace Concurrency;
using namespace Windows::Foundation;
using namespace Windows::Graphics::Printing;
using namespace Windows::Graphics::Printing::OptionDetails;
using namespace Windows::UI::Core;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Media;
using namespace Windows::UI::Xaml::Media::Imaging;
using namespace Windows::UI::Xaml::Printing;

BitmapPrintDocument::BitmapPrintDocument(GetBitmap^ getBitmap)
{
    // Save callback
    this->getBitmap = getBitmap;

    // Element to print
    border = ref new Border();
    image = ref new Image();
    border->Child = image;

    // Get IPrintDocumentSource and attach event handlers
    printDocumentSource = this->DocumentSource;
    this->Paginate += ref new PaginateEventHandler(this, &BitmapPrintDocument::OnPaginate);
    this->GetPreviewPage += ref new GetPreviewPageEventHandler(this, &BitmapPrintDocument::OnGetPreviewPage);
    this->AddPages += ref new AddPagesEventHandler(this, &BitmapPrintDocument::OnAddPages);

    // Attach PrintManager handler
    PrintManager::GetForCurrentView()->PrintTaskRequested += 
        ref new TypedEventHandler<PrintManager^, PrintTaskRequestedEventArgs^>(this, &BitmapPrintDocument::OnPrintManagerPrintTaskRequested);
}

void BitmapPrintDocument::OnPrintManagerPrintTaskRequested(PrintManager^ sender, PrintTaskRequestedEventArgs^ args)
{
    PrintTaskRequestedDeferral^ deferral = args->Request->GetDeferral();

    // Obtain PrintTask
    PrintTask^ printTask = args->Request->CreatePrintTask("Finger Paint", 
                    ref new PrintTaskSourceRequestedHandler(this, &BitmapPrintDocument::OnPrintTaskSourceRequested));

    // Probably set orientation to landscape
    PrintTaskOptionDetails^ optionDetails = 
        PrintTaskOptionDetails::GetFromPrintTaskOptions(printTask->Options);

    PrintOrientationOptionDetails^ orientation =
        dynamic_cast<PrintOrientationOptionDetails^>(optionDetails->Options->Lookup(StandardPrintTaskOptions::Orientation));

    bitmapIsLandscape = false;

    task<void> dispatcherRunTask = create_task(
        this->Dispatcher->RunAsync(CoreDispatcherPriority::Normal, ref new DispatchedHandler([this]()
            {
                BitmapSource^ bitmapSource = getBitmap();

                if (bitmapSource != nullptr)
                    bitmapIsLandscape = bitmapSource->PixelWidth > bitmapSource->PixelHeight;
            })));
    dispatcherRunTask.then([this, orientation, deferral]()
    {
        orientation->TrySetValue(bitmapIsLandscape ? PrintOrientation::Landscape : PrintOrientation::Portrait);
        deferral->Complete();
    });
}

void BitmapPrintDocument::OnPrintTaskSourceRequested(PrintTaskSourceRequestedArgs^ args)
{
    args->SetSource(printDocumentSource);
}

void BitmapPrintDocument::OnPaginate(Object^ sender, PaginateEventArgs^ args)
{
    PrintPageDescription pageDesc = args->PrintTaskOptions->GetPageDescription(0);

    // Get the bitmap
    image->Source = getBitmap();

    // Set padding on the Border
    double left = pageDesc.ImageableRect.Left;
    double top = pageDesc.ImageableRect.Top;
    double right = pageDesc.PageSize.Width - left - pageDesc.ImageableRect.Width;
    double bottom = pageDesc.PageSize.Height - top - pageDesc.ImageableRect.Height;
    border->Padding = ThicknessHelper::FromLengths(left, top, right, bottom);

    this->SetPreviewPageCount(1, PreviewPageCountType::Final);
}

void BitmapPrintDocument::OnGetPreviewPage(Object^ sender, GetPreviewPageEventArgs^ args)
{
    this->SetPreviewPage(args->PageNumber, border);
}

void BitmapPrintDocument::OnAddPages(Object^ sender, AddPagesEventArgs^ args)
{
    this->AddPage(border);
    this->AddPagesComplete();
}

    